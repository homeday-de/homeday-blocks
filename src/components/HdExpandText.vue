<template>
  <div
    :class="{
      'expand-text': true,
      'isExpanded': isExpanded,
    }"
  >
    <div
      ref="wrapper"
      :style="{
        maxHeight: `${wrapperHeight}px`,
      }"
      class="expand-text__wrapper"
    >
      <slot/>
      <p
        ref="sample"
        class="expand-text__wrapper__sample"
      >&nbsp;</p>
    </div>
    <button
      v-if="isToggleVisible"
      class="expand-text__toggle"
      type="button"
      @click="toggleExpanded"
    >
      {{ toggleText }}
      <HdIcon
        :src="chevronIcon"
        transform="rotate(90)"
        class="expand-text__toggle__icon"
      />
    </button>
  </div>
</template>

<script>
import merge from 'lodash/merge';
import HdIcon from 'homeday-blocks/src/components/HdIcon.vue';
import { getMessages } from 'homeday-blocks/src/lang';
import { mediaMatches } from 'homeday-blocks/src/services/breakpoints';
import onResize from 'homeday-blocks/src/services/on-resize';
import { chevronIcon } from 'homeday-blocks/src/assets/small-icons';

export default {
  name: 'HdExpandText',
  components: {
    HdIcon,
  },
  props: {
    // Can be an object with keys set to breakpoints and values the desired
    // lines count
    lines: {
      type: [Number, Object],
      default: 10,
    },
    // If a paragraph has a top margin, it should be specified here
    marginTop: {
      type: Number,
      default: 0,
    },
    // If a paragraph has a bottom margin, it should be specified here
    marginBottom: {
      type: Number,
      default: 0,
    },
    lang: {
      type: String,
      default: 'de',
    },
    texts: {
      type: Object,
      default: () => ({}),
    },
    // Used to approximate the line height before the first mount
    approximateLineHeight: {
      type: Number,
      default: 32,
    },
    // Used if lines is an object and no breakpoint is matched
    linesFallback: {
      type: Number,
      default: 10,
    },
  },
  data() {
    return {
      isExpanded: false,
      isToggleVisible: true,
      wrapperHeight: 0,
      fullHeight: 0,
      chevronIcon,
    };
  },
  computed: {
    t() {
      return merge(getMessages(this.lang), this.texts);
    },
    toggleText() {
      return this.isExpanded
        ? this.t.EXPAND_TEXT.TOGGLE.SHOW_LESS
        : this.t.EXPAND_TEXT.TOGGLE.SHOW_MORE;
    },
  },
  watch: {
    isExpanded: {
      handler() {
        this.resizeWrapper();
      },
      immediate: true,
    },
    lines() {
      this.resizeWrapper();
    },
    linesMobile() {
      this.resizeWrapper();
    },
  },
  mounted() {
    this.$nextTick(this.resizeWrapper);
    this.addResizeEvents();
  },
  beforeDestroy() {
    this.removeResizeEvents();
  },
  methods: {
    // Encapsulated in a method to be able to mock it in the tests
    getScrollHeight(el) {
      return el.scrollHeight;
    },
    toggleExpanded() {
      this.isExpanded = !this.isExpanded;
    },
    getLinesCount() {
      if (typeof this.lines === 'number') {
        return this.lines;
      }

      let linesCount = this.linesFallback;

      Object.keys(this.lines).forEach((breakpoint) => {
        if (!mediaMatches(breakpoint)) {
          return;
        }

        linesCount = this.lines[breakpoint];
      });

      return linesCount;
    },
    calculateFullHeight() {
      if (!this.$refs.wrapper) {
        return;
      }

      this.fullHeight = this.getScrollHeight(this.$refs.wrapper);
    },
    resizeWrapper() {
      this.calculateFullHeight();
      const lines = this.getLinesCount();

      // Check if component is already mounted, if not, we approximate
      // the height
      const lineHeight = this.$refs.sample
        ? this.getScrollHeight(this.$refs.sample)
        : this.approximateLineHeight;
      const linesHeight = (lineHeight * lines) + this.marginTop;

      if (linesHeight + this.marginBottom >= this.fullHeight) {
        this.isToggleVisible = false;
      } else {
        this.isToggleVisible = true;
      }

      if (this.isExpanded && this.$refs.sample) {
        this.wrapperHeight = this.fullHeight;
        return;
      }

      if (linesHeight + this.marginBottom === this.fullHeight) {
        this.wrapperHeight = this.fullHeight + this.marginBottom;
        return;
      }

      this.wrapperHeight = linesHeight;
    },
    addResizeEvents() {
      onResize.onThrottled(this.resizeWrapper);
    },
    removeResizeEvents() {
      onResize.offThrottled(this.resizeWrapper);
    },
  },
};
</script>

<style lang="scss">
@import 'homeday-blocks/src/styles/mixins.scss';

.expand-text {
  $_root: &;
  $_icon-size: 24px;

  &__wrapper {
    position: relative;
    overflow: hidden;
    transition: max-height ($time-s * 2) ease-in-out;

    &__sample {
      position: absolute;
      top: 0;
      pointer-events: none;
      z-index: -1;
      visibility: hidden;

      &::before {
        content: '\2800'
      }
    }
  }

  &__toggle {
    $_root_toggle: &;
    display: flex;
    align-items: center;
    position: relative;
    padding: 0;
    margin-right: $inline-m;
    font-size: 18px;
    line-height: 28px;
    font-family: inherit;
    background-color: transparent;
    border: 0;
    box-shadow: none;
    color: $primary-color;
    cursor: pointer;
    outline: 0;
    transition: color $time-s ease-in-out;

    &:hover,
    &:focus {
      color: getShade($secondary-color, 110);
    }

    @media (min-width: $break-tablet) {
      font-size: 21px;
      line-height: 32px;
    }

    &::before {
      content: '';
      position: absolute;
      bottom: $stack-xs;
      left: 0;
      width: calc(100% - #{$inline-xs + $_icon-size});
      height: 1px;
      background-color: getShade($secondary-color, 110);
    }

    &__icon {
      position: relative;
      width: $_icon-size;
      height: $_icon-size;
      margin-left: $inline-xs;
      transition: transform ($time-s * 2) ease-in-out;

      path {
        fill: currentColor;
      }

      #{$_root}.isExpanded & {
        transform: rotateZ(270deg);
      }
    }
  }
}
</style>
