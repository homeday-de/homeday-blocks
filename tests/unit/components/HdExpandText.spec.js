import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import HdExpandText from '@/components/HdExpandText.vue';

const originalMatchMedia = global.matchMedia;

const wrapperBuilder = wrapperFactoryBuilder(HdExpandText, {
  slots: {
    default: `<p>Look, just because I don't be givin' no man a foot
      massage don't make it right for Marsellus to throw Antwone into
      a glass Muttafu' house, frickin' up the way the talks.
      Muttafu do that shirt to me, he better paralyze my lol, 'cause
      I'll kill the Muttafu, know what I'm sayin'?</p>`,
  },
});

describe('HdExpandText', () => {
  let matchMediaSpy;
  let scrollHeightSpy = jest.fn();

  beforeEach(() => {
    matchMediaSpy = jest.fn(() => ({
      matches: false,
    }));

    global.matchMedia = matchMediaSpy;
  });

  afterEach(() => {
    matchMediaSpy.mockReset();
    global.matchMedia = originalMatchMedia;
    scrollHeightSpy.mockRestore();
  });

  it('renders the component', async () => {
    const wrapper = wrapperBuilder();

    await wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('display partial content', () => {
    const wrapper = wrapperBuilder();

    scrollHeightSpy = jest.spyOn(wrapper.vm, 'getScrollHeight')
      .mockImplementation((el) => {
        if (el === wrapper.vm.$refs.sample) return 32;

        if (el === wrapper.vm.$refs.wrapper) return 5000;

        return 0;
      });

    wrapper.vm.resizeWrapper();

    expect(wrapper.vm.isToggleVisible).toBe(true);
  });

  it('display the full content with no toggle', () => {
    const wrapper = wrapperBuilder();

    scrollHeightSpy = jest.spyOn(wrapper.vm, 'getScrollHeight')
      .mockImplementation((el) => {
        if (el === wrapper.vm.$refs.sample) return 1000;

        if (el === wrapper.vm.$refs.wrapper) return 100;

        return 0;
      });

    wrapper.vm.resizeWrapper();

    expect(wrapper.vm.isToggleVisible).toBe(false);
  });

  it('the content is expandable and shrinkable', async () => {
    const wrapper = wrapperBuilder();

    scrollHeightSpy = jest.spyOn(wrapper.vm, 'getScrollHeight')
      .mockImplementation((el) => {
        if (el === wrapper.vm.$refs.sample) return 32;

        if (el === wrapper.vm.$refs.wrapper) return 5000;

        return 0;
      });

    wrapper.vm.resizeWrapper();
    expect(wrapper.vm.isExpanded).toBe(false);

    await wrapper.vm.$nextTick();

    wrapper.find('.expand-text__toggle').trigger('click');
    expect(wrapper.vm.isExpanded).toBe(true);

    wrapper.find('.expand-text__toggle').trigger('click');
    expect(wrapper.vm.isExpanded).toBe(false);
  });

  it('In the default state, it displays the desidered amount of lines. The vertical margins are considered', () => {
    const marginTop = 72;
    const marginBottom = 61;
    const sampleHeight = 32;
    const lines = 3;
    const wrapper = wrapperBuilder({
      props: {
        lines,
        marginTop,
        marginBottom,
      },
    });

    scrollHeightSpy = jest.spyOn(wrapper.vm, 'getScrollHeight')
      .mockImplementation((el) => {
        if (el === wrapper.vm.$refs.sample) return sampleHeight;

        return 0;
      });

    wrapper.vm.resizeWrapper();

    expect(wrapper.vm.$refs.wrapper.style.maxHeight).toBe(`${(sampleHeight * lines) + marginTop}px`);
  });
});
