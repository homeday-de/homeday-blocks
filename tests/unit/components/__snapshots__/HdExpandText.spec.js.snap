// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdExpandText renders the component 1`] = `
<div class="expand-text">
  <div class="expand-text__wrapper" style="max-height: 0px;">
    <p>Look, just because I don't be givin' no man a foot
      massage don't make it right for Marsellus to throw Antwone into
      a glass Muttafu' house, frickin' up the way the talks.
      Muttafu do that shirt to me, he better paralyze my lol, 'cause
      I'll kill the Muttafu, know what I'm sayin'?</p>
    <p class="expand-text__wrapper__sample">&nbsp;</p>
  </div>
  <!---->
</div>
`;
